### PR TITLE
feat: wire manifest readinessProbe to remote session controller

### PR DESCRIPTION
**Pull request by SalimKayal**
## Summary

The AmaltheaSession CRD already exposes spec.session.readinessProbe, spec.session.port, and spec.session.urlPath. These fields were previously only honored for local sessions (through Kubernetes container probes). This PR forwards the same values into the remote session controller sidecar so its /ready HTTP endpoint reports readiness based on whether the actual remote session is reachable.

## Motivation and Context

For remote sessions managed via FirecREST / RunAI / Slurm, the controller pod was always reporting ready immediately, even if the IDE on the remote compute node had not started yet. We need the Kubernetes readiness state to reflect real user-facing session availability so that ingress and service routing only start once the session is actually usable.

## Changes

*   Injects RSC_SESSION_PORT, RSC_SESSION_URL_PATH, and RSC_READINESS_PROBE_TYPE as environment variables into the remote session controller container (sessionContainerRemote).
*   The sidecar's /ready handler now interprets these values:
    *   none — immediately returns 200 OK.
    *   tcp — attempts a TCP dial to 127.0.0.1:<RSC_SESSION_PORT> (the local wstunnel endpoint); returns 200 on success, 503 otherwise.
    *   http — performs an HTTP GET to 127.0.0.1:<RSC_SESSION_PORT><RSC_SESSION_URL_PATH>; returns 200 on 2xx/3xx responses, 503 otherwise.
*   Fully backward compatible: an empty or unknown probe type, as well as port 0, fall back to the previous behavior of returning 200 OK.

**Tests**

*   Added TestConfigNewFlags covering CLI flags, environment variables, and validation of the readiness probe type.
*   Added TestReadyEndpoint covering all probe modes and edge cases: open/closed TCP ports, HTTP 200/302/500, connection refused, and port-0 fallback.

**Compatibility**

*   No CRD changes required — the fields already exist in amaltheasession_types.go.
*   Existing remote sessions without these values continue to behave exactly as before.

## Notes

- Companion PR: https://github.com/SwissDataScienceCenter/renku-data-services/pull/1350 selects the probe type per session; this PR acts on it in the sidecar. Merge/release together.

### CODE REVIEW

File: internal/remote/server/server.go [hunk 1]
```diff
@@ -114,7 +116,49 @@ func newServer(controller controller.RemoteSessionController) (server *echo.Echo
...
+				return c.NoContent(http.StatusOK)
+			}
+			conn, err := net.DialTimeout("tcp", fmt.Sprintf("127.0.0.1:%d", cfg.SessionPort), 5*time.Second)
+			if err != nil {
+				return c.NoContent(http.StatusServiceUnavailable)
+			}
+			if err := conn.Close(); err != nil {
+				slog.Error("failed to close readiness probe connection", "error", err)
+			}
+			return c.NoContent(http.StatusOK)
```
Comment (olevski) [inline internal/remote/server/server.go:132]: It is better to use the request context for this. That way if the request is cancelled the tcp connection is also cancelled. And we can still have the timeout we want.

```suggestion
		case string(amaltheadevv1alpha1.TCP):
			if cfg.SessionPort == 0 {
				return c.NoContent(http.StatusOK)
			}
            dialer := net.Dialer{}
			dialCtx, dialCtxCancel := context.WithTimeout(c.Request().Context(), 5*time.Second)
			defer dialCtxCancel()
			conn, err := dialer.DialContext(dialCtx, "tcp", fmt.Sprintf("127.0.0.1:%d", cfg.SessionPort))

			if err != nil {
				return c.NoContent(http.StatusServiceUnavailable)
			}
			if err := conn.Close(); err != nil {
				slog.Error("failed to close readiness probe connection", "error", err)
			}
			return c.NoContent(http.StatusOK)
```

File: internal/remote/server/server.go [hunk 2]
```diff
@@ -114,7 +116,49 @@ func newServer(controller controller.RemoteSessionController) (server *echo.Echo
...
+			}
+			defer func() {
+				if err := resp.Body.Close(); err != nil {
+					slog.Error("failed to close readiness probe response body", "error", err)
+				}
+			}()
+			if resp.StatusCode >= 200 && resp.StatusCode < 400 {
+				return c.NoContent(http.StatusOK)
+			}
+			return c.NoContent(http.StatusServiceUnavailable)
```
Comment (olevski) [inline internal/remote/server/server.go:157]: Same as above, the request for the health check should be a child of the request to the server.

```suggestion
		case string(amaltheadevv1alpha1.HTTP):
			if cfg.SessionPort == 0 {
				return c.NoContent(http.StatusOK)
			}
			client := &http.Client{
				Timeout: 5 * time.Second,
				CheckRedirect: func(req *http.Request, via []*http.Request) error {
					return http.ErrUseLastResponse
				},
			}
			url := fmt.Sprintf("http://127.0.0.1:%d%s", cfg.SessionPort, cfg.SessionURLPath)
			req, err := http.NewRequestWithContext(c.Request().Context(), "GET", url, nil)
			if err != nil {
				return c.NoContent(http.StatusServiceUnavailable)
			}
			resp, err := client.Do(req)
			if err != nil {
				return c.NoContent(http.StatusServiceUnavailable)
			}
			defer func() {
				if err := resp.Body.Close(); err != nil {
					slog.Error("failed to close readiness probe response body", "error", err)
				}
			}()
			if resp.StatusCode >= 200 && resp.StatusCode < 400 {
				return c.NoContent(http.StatusOK)
			}
			return c.NoContent(http.StatusServiceUnavailable)
```

File: internal/remote/server/server_test.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+			}(),
+			setupBackend: func() func() {
+				ln, err := net.Listen("tcp", "127.0.0.1:18000")
+				if err != nil {
+					panic(err)
+				}
+				return func() { _ = ln.Close() }
+			},
+			wantStatus: http.StatusOK,
+		},
```
Comment (olevski) [inline internal/remote/server/server_test.go:83]: This is a way to not hardcode the port to run the test server on.


```
func getFreePortOrDie() int {
	var a *net.TCPAddr
	var err error
	if a, err = net.ResolveTCPAddr("tcp", "localhost:0"); err == nil {
		var l *net.TCPListener
		if l, err = net.ListenTCP("tcp", a); err == nil {
			defer func() {
				err := l.Close()
				if err != nil {
					panic(err)
				}
			}()
			return l.Addr().(*net.TCPAddr).Port
		}
		panic(err)
	}
	panic(err)
}
```

From https://github.com/SwissDataScienceCenter/renku-frontend-buildpacks/blob/main/tests/e2e/utils.go#L162C1-L179C2

**Review by olevski — DISMISSED** [review]
Thanks Salim. Just a few minor things. You can ignore the note about finding a free port if you wish.

File: internal/remote/config/config.go [hunk 4]
```diff
@@ -72,6 +86,30 @@ func SetFlags(cmd *cobra.Command) error {
...
+
+	cmd.Flags().String(sessionURLPathFlag, "/", "URL path for the HTTP readiness probe")
+	if err := viper.BindPFlag(sessionURLPathFlag, cmd.Flags().Lookup(sessionURLPathFlag)); err != nil {
+		return err
+	}
+	if err := viper.BindEnv(sessionURLPathFlag, configUtils.AsEnvVarFlag(sessionURLPathFlag)); err != nil {
+		return err
+	}
+
+	cmd.Flags().String(readinessProbeTypeFlag, "", "readiness probe type: none, tcp, or http")
```
Comment (leafty) [inline internal/remote/config/config.go]: Why is the default `""` and not `"none"`?

File: internal/remote/server/server.go [hunk 5]
```diff
@@ -114,7 +116,49 @@ func newServer(controller controller.RemoteSessionController) (server *echo.Echo
...
 	// Readiness endpoint
 	e.GET("/ready", func(c echo.Context) error {
-		return c.NoContent(http.StatusOK)
+		switch cfg.ReadinessProbeType {
+		case string(amaltheadevv1alpha1.None):
+			return c.NoContent(http.StatusOK)
+		case string(amaltheadevv1alpha1.TCP):
+			if cfg.SessionPort == 0 {
+				return c.NoContent(http.StatusOK)
+			}
```
Comment (leafty) [inline internal/remote/server/server.go]: This should not happen in my opinion: using a port value of `0` is invalid. The default values should map to `amaltheadevv1alpha1.None`.

File: internal/remote/server/server.go [hunk 6]
```diff
@@ -114,7 +116,49 @@ func newServer(controller controller.RemoteSessionController) (server *echo.Echo
 
 	// Readiness endpoint
 	e.GET("/ready", func(c echo.Context) error {
-		return c.NoContent(http.StatusOK)
+		switch cfg.ReadinessProbeType {
+		case string(amaltheadevv1alpha1.None):
+			return c.NoContent(http.StatusOK)
```
Comment (leafty) [inline internal/remote/server/server.go]: Since we also have a `default` case, you can remove the `None` case.

File: internal/remote/config/config.go [hunk 7]
```diff
@@ -97,11 +135,21 @@ func GetConfig() (cfg RemoteSessionControllerConfig, err error) {
...
 	return cfg, nil
 }
 
 func (cfg *RemoteSessionControllerConfig) Validate() error {
+	if cfg.ReadinessProbeType != "" &&
+		cfg.ReadinessProbeType != string(amaltheadevv1alpha1.None) &&
+		cfg.ReadinessProbeType != string(amaltheadevv1alpha1.TCP) &&
+		cfg.ReadinessProbeType != string(amaltheadevv1alpha1.HTTP) {
+		return fmt.Errorf("invalid readiness probe type: %s", cfg.ReadinessProbeType)
+	}
```
Comment (leafty) [inline internal/remote/config/config.go:150]: Is there any reason to not update the configuration to have `none` as the default? This way you can remove the defensive code in the endpoint implementation.

File: internal/remote/server/server_test.go (hunk 3 above)
Comment (SalimKayal) [reply internal/remote/server/server_test.go:83]: thanks, added

File: internal/remote/config/config.go (hunk 7 above)
Comment (SalimKayal) [reply internal/remote/config/config.go:150]: added default as none

File: internal/remote/config/config.go (hunk 4 above)
Comment (SalimKayal) [reply internal/remote/config/config.go]: changed

File: internal/remote/server/server.go (hunk 5 above)
Comment (SalimKayal) [reply internal/remote/server/server.go]: removed

File: internal/remote/server/server.go (hunk 6 above)
Comment (SalimKayal) [reply internal/remote/server/server.go]: removed

File: internal/remote/server/server.go [hunk 8]
```diff
@@ -114,7 +116,52 @@ func newServer(controller controller.RemoteSessionController) (server *echo.Echo
...
+				return c.NoContent(http.StatusServiceUnavailable)
+			}
+			if err := conn.Close(); err != nil {
+				slog.Error("failed to close readiness probe connection", "error", err)
+			}
+			return c.NoContent(http.StatusOK)
+		case string(amaltheadevv1alpha1.HTTP):
+			if cfg.SessionPort == 0 {
+				return c.NoContent(http.StatusOK)
+			}
```
Comment (leafty) [inline internal/remote/server/server.go]:
```suggestion
```

File: internal/remote/server/server.go [hunk 9]
```diff
@@ -114,7 +116,52 @@ func newServer(controller controller.RemoteSessionController) (server *echo.Echo
...
+				if err := resp.Body.Close(); err != nil {
+					slog.Error("failed to close readiness probe response body", "error", err)
+				}
+			}()
+			if resp.StatusCode >= 200 && resp.StatusCode < 400 {
+				return c.NoContent(http.StatusOK)
+			}
+			return c.NoContent(http.StatusServiceUnavailable)
+		default:
+			// Unconfigured / unknown: preserve old behavior for backward compatibility
```
Comment (leafty) [inline internal/remote/server/server.go]:
```suggestion
			// Case None or unset: preserve old behavior for backward compatibility
```

**Review by leafty — APPROVED** [review]
